arraycontext/container/traversal.py

```diff
@@ -533,6 +533,9 @@ def freeze(
 
     See :meth:`ArrayContext.thaw`.
     """
+    if actx is not None and actx.supports_freezing_containers:
```

Comment:
We should just deprecate these then (and expect all array contexts to support container freezes).

Reply:
Given that the freeze logic is specific to the containers (freeze is a single-dispatch function), deprecating this won't be possible.

Reply:
This should also be added to [meshmode.dof_array](https://github.com/inducer/meshmode/blob/b795b77d49422da2da90f1c0b07668d2723ae048/meshmode/dof_array.py#L381-L389) then?

Reply:
As a side note, I really don't recall why we made this `singledispatch`-ed and can't think of a reason for it now. Anyone remember something?

Reply:
>  I really don't recall why we made this singledispatch-ed and can't think of a reason for it now. Anyone remember something?

IIUC, array container initialization might be specific to the container type, and instead of forcing any constructor API we ask the container's writer to also provide a freeze. For example, DOFArray takes in a `.array_context` argument during initialization that must be set to `None` on getting frozen.

> This should also be added to [meshmode.dof_array](https://github.com/inducer/meshmode/blob/b795b77d49422da2da90f1c0b07668d2723ae048/meshmode/dof_array.py#L381-L389) then?

Yep!

Reply:
> IIUC, array container initialization might be specific to the container type, and instead of forcing any constructor API we ask the container's writer to also provide a freeze. For example, DOFArray takes in a `.array_context` argument during initialization that must be set to `None` on getting frozen.

And using `deserialize_container` doesn't work because the template arg isn't necessarily frozen. Hm.. annoying..

Thanks for the reminder!

Reply:
Revisiting this: given the new logic, I think we ought to just deprecate the stand-alone `freeze` in favor of `actx.freeze`. (same for `thaw`) I don't see a strong rationale for keeping them.

Reply:
In view of #162, I'd prefer to reduce our dependence on these, so deprecating them now seems like a good move. I know that this is will be super noisy on the application end, but I think it's a necessary (if painful) step.

I'd welcome feedback though. (cc @mtcam)